### PR TITLE
Fix nested label formatting: Restore the original behavior of not app…

### PR DESCRIPTION
**Pull request by ksooo**
## Description

Fix nested label formatting: Restore the original behavior of not applying style when a closing tag is unmatched

## Motivation and context

Fix a regression introduced in #28280

## How has this been tested?

Manual testcase.

## What is the effect on users?

Text with unmatched format tags looks the same as before #28280.

## Screenshots (if appropriate):
n/a

## Types of change

- [X] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **Student submission** (PR was done for educational purposes and will be treated as such)
- [ ] **None of the above** (please explain below)

## Checklist:

- [X] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [X] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [] I have added tests to cover my change
- [X] All new and existing tests passed

### CODE REVIEW

**Comment by CrystalP** [comment]
Makes sense to check more strictly.

I didn't runtime test, but wouldn't this unbalanced string pass?
[B]text1[B]text2[/B]
I'm not sure what should happen, can't tell which [B] the only [/B] should match.

I think a test with opening and closing in different order would be good as it's expected to pass right?
[B]A[I]B[/B]C[/I]

Wouldn't it be good to log the unbalanced format tags? I imagine you noticed the behavior change because of a buggy skin/addon? As those situations should not happen, or did I miss something?

Is there adequate quoting provided for the exceptional cases where a format tag is part of the text to be displayed and should appear without interpretation? (beyond the PR's scope I suppose)

File: xbmc/guilib/GUITextLayout.cpp [hunk 1]
```diff
@@ -417,20 +417,22 @@ void CGUITextLayout::ParseText(const std::wstring& text,
     { // bold
       pos += 2;
       if (on)
-        ++boldDepth;
+      {
+        if (text.find(L"[/B]", pos) != std::wstring::npos)
```
Comment (garbear) [inline xbmc/guilib/GUITextLayout.cpp:421]: I think this would fail for text like `[B][B]unbold[B]bold[/B]`

**Comment by ksooo** [comment]
@CrystalP @garbear thanks for the review. For sure this code is far away from being perfect. Intention for this PR was not to fix all parsing issues (by writing a real parser), only to restore the old behavior to not apply the style if no closing tag was found for an opening tag. I hope I achieved exactly this. That is all I wanted. On top I added a bunch of tests. Everything else is for another PR. I hope that is fair enough.

File: xbmc/guilib/GUITextLayout.cpp (hunk 1 above)
Comment (ksooo) [reply xbmc/guilib/GUITextLayout.cpp:421]: What do you mean?

File: xbmc/guilib/test/TestGUITextLayoutNestedFormatting.cpp [hunk 2]
```diff
@@ -0,0 +1,290 @@
...
+
+  // After italic closing, still bold
+  EXPECT_TRUE(HasStyleAtPosition(text, FONT_STYLE_BOLD, 20));
+}
+
+/*!
+ * Test: Opening tag without matching closing tag should not apply style
+ * 
+ * If an opening [B] has no matching [/B], per the fix, the depth should not increment
+ */
```
Comment (CrystalP) [inline xbmc/guilib/test/TestGUITextLayoutNestedFormatting.cpp]: Line to be removed, if it had to be somewhere it would be in the commit message or the PR description.

File: xbmc/guilib/test/TestGUITextLayoutNestedFormatting.cpp [hunk 3]
```diff
@@ -0,0 +1,290 @@
...
+  void SetUp() override
+  {
+    // Placeholder for any setup needed for tests
+  }
+
+  void TearDown() override
+  {
+    // Placeholder for any cleanup needed
+  }
+
```
Comment (CrystalP) [inline xbmc/guilib/test/TestGUITextLayoutNestedFormatting.cpp]: The functions do nothing, please remove.

File: xbmc/guilib/test/TestGUITextLayoutNestedFormatting.cpp [hunk 4]
```diff
@@ -0,0 +1,290 @@
...
+};
+
+/*!
+ * Test: Matched bold tags should apply bold style
+ */
+TEST_F(TestGUITextLayoutNestedFormatting, MatchedBoldTagsApplyStyle)
+{
+  std::wstring text = L"[B]bold text[/B]";
+
+  // Text "bold text" should have bold style applied
```
Comment (CrystalP) [inline xbmc/guilib/test/TestGUITextLayoutNestedFormatting.cpp]: Most (all?) of the in-function comments are unnecessary and could be removed.

**Review by CrystalP — COMMENTED** [review]
Unnecessary things from the AI generated unit tests to be removed.

**Comment by ksooo** [comment]
Unneeded comments from AI-generated tests are now removed.

File: xbmc/guilib/GUITextLayout.cpp (hunk 1 above)
Comment (garbear) [reply xbmc/guilib/GUITextLayout.cpp:421]: What @CrystalP said:

> I didn't runtime test, but wouldn't this unbalanced string pass? [B]text1[B]text2[/B]

But I see your reason for wanting restoration of the simple behavior without handling all sorts of edge cases.

File: xbmc/guilib/test/TestGUITextLayoutNestedFormatting.cpp [hunk 5]
```diff
@@ -0,0 +1,246 @@
...
+#include "guilib/GUIFont.h"
+#include "guilib/GUITextLayout.h"
+
+#include <string>
+#include <vector>
+
+#include <gtest/gtest.h>
+
+/*!
+ * @brief Test suite for GUITextLayout nested label formatting *
```
Comment (CrystalP) [inline xbmc/guilib/test/TestGUITextLayoutNestedFormatting.cpp]: style: the guidelines ask for \ not @ and there is a stray * at the end of the line.

**Comment by CrystalP** [comment]
That's better. I'd say that HasStyleAtPosition does too much and the parsing and testing should be separated to avoid redundant parsing of the same string (ex. NestedMatchedBoldTags). Parametrized tests would probably be best to avoid repetitions.

**Comment by ksooo** [comment]
I think i spent more than enough time with this PR. I only wanted to fix a regression introduced by somebody else. Those tests are already a bonus. I can remove them if people feel they are still not good enough.

**Comment by ksooo** [comment]
Tests removed.
